src/tapedfunction.jl

```diff
@@ -56,72 +25,132 @@ mutable struct TapedFunction{F}
         end
 
         ir = CodeInfoTools.code_inferred(f, args_type...)
-        tf = new{F}() # leave some fields to be undef
-        tf.func, tf.arity, tf.ir = f, length(args), ir
-        tf.tape = RawTape()
-        tf.counter = 1
+        tape = RawTape()
+        utape = Vector{FunctionWrapper{Nothing, Tuple{TapedFunction}}}()
+        bindings = translate!(tape, ir)
 
-        translate!(tf, ir)
+        tf = new{F}(f, length(args), ir, tape, utape, 1, bindings, :none)
         TRCache[cache_key] = tf # set cache
+        # unify!(tf)
         return tf
     end
 
     function TapedFunction(tf::TapedFunction{F}) where {F}
-        new{F}(tf.func, tf.arity, tf.ir, tf.tape,
+        new{F}(tf.func, tf.arity, tf.ir, tf.tape, tf.unified_tape,
                tf.counter, tf.bindings, :none)
     end
 end
 
 const TRCache = LRU{Tuple, TapedFunction}(maxsize=10)
 
+function unify!(tf::TapedFunction)
+    length(tf.tape) == length(tf.unified_tape) && return
+    for ins in tf.tape
+        push!(tf.unified_tape, FunctionWrapper{Nothing, Tuple{TapedFunction}}(ins))
+    end
+end
+
+# const TypedFunction = FunctionWrapper
+struct TypedFunction{OT, IT<:Tuple}
+    func::Function
+    retval::Ref{OT}
+    TypedFunction{OT, IT}(f::Function) where {OT, IT<:Tuple} = new{OT, IT}(f, Ref{OT}())
+end
+function (f::TypedFunction{OT, IT})(args...) where {OT, IT<:Tuple}
+    output = f.func(args...)
+    OT === Nothing ? (f.retval[] = nothing) : (f.retval[] = output)
+    return f.retval[]
```

Comment:
A bit simpler, without accessing `retval` twice:
```suggestion
    f.reval[] = OT === Nothing ? nothing : output
```